### ISSUE TITLE
Massive RX packet loss on Mellanox cards (mlx5)

### ISSUE THREAD

**Issue by driesdewinter** [luigirizzo/netmap#764]
The high level behavior that I observe when putting a Mellanox network port in netmap mode, is that the majority of input packets is lost. I can only receive a very low bitrate in my application (order of magnitude a few Mbps). Apparently it is because after every `ioctl(...RXSYNC...)`-call, I only get at most one packet, as if the RX ring size was just one packet. However, I set the ring size to 4096 via ethtool and the ring's `num_slots` is really 4096. Also, the same application code works well with Intel 10G ports (i40e).

So I added some extra traces in `mlx5e_netmap_rxsync()` to see what is going on there.
The first time it is called,
- the first part (reclaim buffers) moves `kring->nr_hwcur` from 4095 to 0. So one extra buffer is made available for reception, after the first 4095 buffers were already preserved by `mlx5e_netmap_configure_rx_ring()`.
- the second part (import newly received packets) doesn't do anything, `kring->nr_hwtail` is 0 and stays 0.

The second time it is called,
- the first part does nothing (`kring->nr_hwcur: 0 -> 0`), so no buffers reclaimed.
- the second part imports a single packet (`kring->nr_hwtail: 0 -> 1`)

The 3rd time I get `kring->nr_hwcur: 0 -> 1; kring->nr_hwtail: 1 -> 1`.
The 4th time I get  `kring->nr_hwcur: 1 -> 1; kring->nr_hwtail: 1 -> 2`.
And so on. This pattern is repeated infinitely.

It seems that there is something wrong with the initial buffers preserved by `mlx5e_netmap_configure_rx_ring()` and that the only slot that is really available for receiving packets is that one extra slot that is released by `mlx5e_netmap_rxsync()`. So as a quick and dirty workaround I made the following patch:
```
diff --git a/LINUX/mlx5_netmap_linux.h b/LINUX/mlx5_netmap_linux.h
index 2170068e..fb054955 100644
--- a/LINUX/mlx5_netmap_linux.h
+++ b/LINUX/mlx5_netmap_linux.h
@@ -648,7 +648,7 @@ int mlx5e_netmap_configure_rx_ring(struct mlx5e_rq *rq, int ring_nr) {
   if (!slot)
     return 0; /* not in native netmap mode */

-  lim = na->num_rx_desc - 1 - nm_kr_rxspace(na->rx_rings[ring_nr]);
+  lim = 1; //na->num_rx_desc - 1 - nm_kr_rxspace(na->rx_rings[ring_nr]);

   while (!mlx5_wq_cyc_is_full(wq) && (count < lim)) {
```
So actually just disabling the preservation of buffers largely. With this patch, it works fine for me. The first time `mlx5e_netmap_rxsync()` is called, 4095 buffers are released instead of just one and they are all really used for receiving packets. But of course this cannot be considered as a real fix. So I wonder: am I doing something wrong? Or did I hit a real bug that needs a proper fix? Anyway, without this hack I cannot get it to work well so I must be doing something differently that I don't understand.

Some extra context info:
I'm working on a CentOS server, more precisely `CentOS Linux release 7.9.2009 (Core)`
Kernel version is `3.10.0-1160.el7.x86_64`
My original mlx5_core driver was version 5.0-0, so I configured netmap as follows:
`../configure --drivers=mlx5 --select-version=mlx5:5.0-1.0.0.0`
The NIC is of type Mellanox Technologies MT27800 Family [ConnectX-5].

**Comment by kierank** [comment]
Does this happen on 4.5 or 4.6?

**Comment by driesdewinter** [comment]
What do you mean with 4.5 or 4.6? My kernel version is 3.10 and the mlx5 driver version is 5.0.

In the meantime I figured out that my quick fix also doesn't work. It worked initially because I have a specific payload on the wire that only consists of fixed length packets. But as soon as I get packets with variable length, I notice that the packet lengths in the netmap rings are not correct: the packet lengths in the netmap ring do not match with the payload pointed to. So it seems there is a discrepancy going on between the wq/cq bookkeeping in the driver and the ring bookkeeping in netmap. I still believe it has something to do with those buffers prepopulated by mlx5e_netmap_configure_rx_ring(), although I can't see what is wrong with it. Is it possible that the work queues are still accessed by some other code path although an interface is put in netmap mode under certain circumstances, which would confuse the netmap ring bookkeeping?

**Comment by kierank** [comment]
Try with Mellanox driver 4.5 or 4.6 so we can see which driver version broke it.

**Comment by driesdewinter** [comment]
I tried to test with driver versions 4.5 and 4.6 but the driver didn't compile in my environment (the initial driver code, so without the patches). Perhaps simple to fix, but I didn't get to the bottom of this.

Anyway, I think I know what broke it: https://github.com/torvalds/linux/commit/5ee090ed0da649b1febae2b7c285ac77d1e55a0c and more specifically the function `mlx5_wq_cyc_reset()` which gets called indirectly from `mlx5e_open_rq()` (via `mlx5e_modify_rq_state()`). This resets the work queue after `mlx5e_netmap_configure_rx_ring()` was called, which perfectly explains the behavior that I'm observing. 

So then the question becomes what is the right fix? I tried to move the call `mlx5e_netmap_configure_rx_ring()` from the end of `mlx5e_alloc_rq()` to the end of `mlx5e_open_rq()` and that works for me. See also https://github.com/driesdewinter/netmap/commit/f14770c7f84985fccd7a3f28b5ccf167f38ac587 . Do you see any problems with that? Perhaps that call to `mlx5e_netmap_configure_rx_ring()` was placed at the end of `mlx5e_alloc_rq()` for a very good reason that is now broken?

**Comment by kierank** [comment]
I personally think it's ok. 

@JDarnley Can you test this change on our side?

**Comment by JDarnley** [comment]
That commit does work for us and our software.  I'm going to try to copy it over to 5.1 and 5.2 and test.

**Comment by JDarnley** [comment]
@driesdewinter The change works on both 5.1 and 5.2 (from quick testing).  Do you want to make a PR for the commit you have above?  If not then I can make one to fix both 5.0 and 5.1 together.
